File: apps/band/src/common/render-protocol.js
const LIMITS = Object.freeze({
  envelopeBytes: 1024,
  payloadBytes: 8192,
  width: 212,
  height: 520,
  maximumPrimitives: 0,
  formatVersion: 1,
  maximumIdentifierBytes: 24
})

const REJECT_CODES = Object.freeze([
  "unsupportedRenderer",
  "unsupportedFormatVersion",
  "busy",
  "payloadTooLarge",
  "tooManyPrimitives",
  "invalidDimensions",
  "insufficientStorage"
])

function utf8Length(value) {
  if (typeof value !== "string") return 0
  let length = 0
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index)
    if (code < 128) length += 1
    else if (code < 2048) length += 2
    else if (code >= 55296 && code <= 56319 && index + 1 < value.length) {
      index += 1
      length += 4
    } else length += 3
  }
  return length
}

function validIdentifier(value) {
  return typeof value === "string" && utf8Length(value) >= 1 &&
    utf8Length(value) <= LIMITS.maximumIdentifierBytes && /^[\x20-\x7e]+$/.test(value)
}

function validSHA256(value) {
  return typeof value === "string" && /^[0-9a-f]{64}$/.test(value)
}

function validInteger(value) {
  return Number.isInteger(value) && Number.isFinite(value)
}

function reject(code) {
  return { ok: false, code }
}

function safePixel(x, y) {
  if (x < 12 || x > 200 || y < 12 || y > 508) return false
  if (y < 106) return Math.hypot(x - 106, y - 106) <= 94
  if (y > 413) return Math.hypot(x - 106, y - 413) <= 94
  return true
}

function safeCenter(x, y, width, height, margin = 6) {
  const halfWidth = width / 2 + margin, halfHeight = height / 2 + margin
  return safePixel(x - halfWidth, y - halfHeight) && safePixel(x + halfWidth, y - halfHeight) &&
    safePixel(x - halfWidth, y + halfHeight) && safePixel(x + halfWidth, y + halfHeight)
}

function validPreview(preview) {
  return preview && typeof preview === "object" && !Array.isArray(preview) &&
    ["straight", "left", "right", "uTurn", "roundabout", "arrive"].includes(preview.maneuver) &&
    validInteger(preview.distanceM) && preview.distanceM >= 0 &&
    typeof preview.street === "string" && utf8Length(preview.street) <= 48 &&
    validInteger(preview.x) && validInteger(preview.y) && safeCenter(preview.x, preview.y, 46, 54) &&
    validInteger(preview.heading) && preview.heading >= 0 && preview.heading < 8 &&
    ["visible", "edge", "hidden"].includes(preview.destinationMode) &&
    validInteger(preview.destinationX) && validInteger(preview.destinationY) &&
    (preview.destinationMode === "hidden" ? preview.destinationX === 0 && preview.destinationY === 0 :
      safeCenter(preview.destinationX, preview.destinationY, 20,
        preview.destinationMode === "visible" ? 24 : 20, preview.destinationMode === "edge" ? 0 : 6))
}

function validatePrepare(body, options = {}) {
  if (options.prepared) return reject("busy")
  if (!body || typeof body !== "object" || Array.isArray(body)) return reject("unsupportedRenderer")
  if (body.renderer !== "raster") return reject("unsupportedRenderer")
  if (body.formatVersion !== LIMITS.formatVersion) return reject("unsupportedFormatVersion")
  if (body.format !== "image/png") {
    return reject("unsupportedFormatVersion")
  }
  if (body.width !== LIMITS.width || body.height !== LIMITS.height) return reject("invalidDimensions")
  if (!validInteger(body.bytes) || body.bytes <= 0 || body.bytes > LIMITS.payloadBytes) return reject("payloadTooLarge")
  if (!validInteger(body.primitives) || body.primitives < 0 || body.primitives > LIMITS.maximumPrimitives) {
    return reject("tooManyPrimitives")
  }
  if (!validIdentifier(body.runId) || !validIdentifier(body.sceneId) || !validSHA256(body.sha256)) {
    return reject("unsupportedFormatVersion")
  }
  if (options.availableStorageBytes !== undefined && body.bytes > options.availableStorageBytes) {
    return reject("insufficientStorage")
  }
  if (body.preview !== undefined && !validPreview(body.preview)) return reject("unsupportedFormatVersion")
  return {
    ok: true,
    prepared: {
      runId: body.runId,
      sceneId: body.sceneId,
      renderer: body.renderer,
      format: body.format,
      formatVersion: body.formatVersion,
      width: body.width,
      height: body.height,
      bytes: body.bytes,
      sha256: body.sha256,
      primitives: body.primitives,
      preview: body.preview
    }
  }
}

function validateAssetBegin(body, prepared) {
  if (!prepared) return reject("notPrepared")
  if (!body || typeof body !== "object" ||
    body.run !== prepared.runId || body.scene !== prepared.sceneId ||
    body.renderer !== prepared.renderer || body.formatVersion !== prepared.formatVersion ||
    body.width !== prepared.width || body.height !== prepared.height ||
    body.bytes !== prepared.bytes || body.sha256 !== prepared.sha256 ||
    body.primitives !== prepared.primitives || body.format !== prepared.format) {
    return reject("notPrepared")
  }
  return { ok: true }
}

export { LIMITS, REJECT_CODES, validIdentifier, validSHA256, validatePrepare, validateAssetBegin }
export default { LIMITS, REJECT_CODES, validIdentifier, validSHA256, validatePrepare, validateAssetBegin }

File: packages/BlueBandKit/Sources/BlueBandMapCore/RenderProtocol.swift
import Foundation
import BlueBandCore

public enum RenderKind: String, Codable, CaseIterable, Sendable {
    case raster
}

public enum RenderFormat: String, Codable, Sendable {
    case raster = "image/png"
}

public enum RenderRejectCode: String, Codable, CaseIterable, Sendable {
    case unsupportedRenderer
    case unsupportedFormatVersion
    case busy
    case payloadTooLarge
    case tooManyPrimitives
    case invalidDimensions
    case insufficientStorage
}

public enum RenderProtocolError: Swift.Error, Equatable, Sendable {
    case invalidIdentifier
    case invalidDimensions
    case invalidFormat
    case unsupportedFormatVersion
    case emptyPayload
    case payloadTooLarge
    case tooManyPrimitives
    case invalidSHA256
    case invalidMilliseconds
    case invalidByteCount
}

public enum RenderProtocol {
    public static let prepareTopic = "render.prepare"
    public static let readyTopic = "render.ready"
    public static let rejectTopic = "render.reject"
    public static let resultTopic = "render.result"

    public static let formatVersion = 1
    public static let viewportWidth = 212
    public static let viewportHeight = 520
    public static let maximumPayloadBytes = 8_192
    public static let maximumPrimitives = 0
    public static let maximumIdentifierBytes = 24

    public static func isValidIdentifier(_ value: String) -> Bool {
        let bytes = value.utf8
        return (1...maximumIdentifierBytes).contains(bytes.count)
            && bytes.allSatisfy { (0x20...0x7E).contains($0) }
    }

    public static func isValidSHA256(_ value: String) -> Bool {
        value.utf8.count == 64 && value.utf8.allSatisfy {
            (0x30...0x39).contains($0) || (0x61...0x66).contains($0)
        }
    }

    static func validate(
        runID: String,
        sceneID: String,
        renderer: RenderKind,
        format: String,
        formatVersion: Int,
        width: Int,
        height: Int,
        bytes: Int,
        sha256: String,
        primitives: Int
    ) throws {
        guard isValidIdentifier(runID), isValidIdentifier(sceneID) else {
            throw RenderProtocolError.invalidIdentifier
        }
        guard width == viewportWidth, height == viewportHeight else {
            throw RenderProtocolError.invalidDimensions
        }
        guard renderer == .raster, format == RenderFormat.raster.rawValue else {
            throw RenderProtocolError.invalidFormat
        }
        guard formatVersion == Self.formatVersion else {
            throw RenderProtocolError.unsupportedFormatVersion
        }
        guard (1...maximumPayloadBytes).contains(bytes) else {
            throw bytes <= 0 ? RenderProtocolError.emptyPayload : RenderProtocolError.payloadTooLarge
        }
        guard (0...maximumPrimitives).contains(primitives) else {
            throw RenderProtocolError.tooManyPrimitives
        }
        guard isValidSHA256(sha256) else {
            throw RenderProtocolError.invalidSHA256
        }
    }
}

public struct RenderNavigationPreview: Equatable, Codable, Sendable {
    public enum Error: Swift.Error, Equatable, Sendable {
        case invalidDistance, invalidMarker, invalidHeading, invalidDestination
    }

    public let maneuver: NavigationManeuver
    public let distanceMeters: Int
    public let street: String
    public let x: Int
    public let y: Int
    public let headingBucket: Int
    public let destinationMode: DestinationPresentationMode
    public let destinationX: Int
    public let destinationY: Int

    public init(
        maneuver: NavigationManeuver,
        distanceMeters: Int,
        street: String,
        x: Int,
        y: Int,
        headingBucket: Int,
        destinationMode: DestinationPresentationMode,
        destinationX: Int,
        destinationY: Int
    ) throws {
        guard distanceMeters >= 0 else { throw Error.invalidDistance }
        guard BandDisplaySafeMask.smartBand10PhotoEstimate.contains(
            center: ScreenPoint(x: x, y: y),
            resourceWidth: 46,
            resourceHeight: 54
        ) else { throw Error.invalidMarker }
        guard (0..<8).contains(headingBucket) else { throw Error.invalidHeading }
        if destinationMode == .hidden {
            guard destinationX == 0, destinationY == 0 else { throw Error.invalidDestination }
        } else {
            let height = destinationMode == .visible ? 24 : 20
            let mask = destinationMode == .edge
                ? BandDisplaySafeMask.smartBand10PhotoEstimate.withoutVisualMargin
                : BandDisplaySafeMask.smartBand10PhotoEstimate
            guard mask.contains(
                center: ScreenPoint(x: destinationX, y: destinationY),
                resourceWidth: 20,
                resourceHeight: height
            ) else { throw Error.invalidDestination }
        }
        var boundedStreet = street
        while boundedStreet.utf8.count > 48 { boundedStreet.removeLast() }
        self.maneuver = maneuver
        self.distanceMeters = distanceMeters
        self.street = boundedStreet
        self.x = x
        self.y = y
        self.headingBucket = headingBucket
        self.destinationMode = destinationMode
        self.destinationX = destinationX
        self.destinationY = destinationY
    }

    public func jsonBody() -> [String: JSONValue] {
        [
            "maneuver": .string(maneuver.rawValue),
            "distanceM": .number(Double(distanceMeters)),
            "street": .string(street),
            "x": .number(Double(x)),
            "y": .number(Double(y)),
            "heading": .number(Double(headingBucket)),
            "destinationMode": .string(destinationMode.rawValue),
            "destinationX": .number(Double(destinationX)),
            "destinationY": .number(Double(destinationY)),
        ]
    }
}

public struct RenderPrepareBody: Equatable, Codable, Sendable {
    public let runID: String
    public let sceneID: String
    public let renderer: RenderKind
    public let format: String
    public let formatVersion: Int
    public let width: Int
    public let height: Int
    public let bytes: Int
    public let sha256: String
    public let primitives: Int
    public let preview: RenderNavigationPreview?

    public init(
        runID: String,
        sceneID: String,
        asset: RenderAsset,
        preview: RenderNavigationPreview? = nil
    ) throws {
        try RenderProtocol.validate(
            runID: runID,
            sceneID: sceneID,
            renderer: asset.kind,
            format: asset.format.rawValue,
            formatVersion: asset.formatVersion,
            width: asset.width,
            height: asset.height,
            bytes: asset.byteCount,
            sha256: asset.sha256,
            primitives: asset.primitives
        )
        self.runID = runID
        self.sceneID = sceneID
        self.renderer = asset.kind
        self.format = asset.format.rawValue
        self.formatVersion = asset.formatVersion
        self.width = asset.width
        self.height = asset.height
        self.bytes = asset.byteCount
        self.sha256 = asset.sha256
        self.primitives = asset.primitives
        self.preview = preview
    }

    public func jsonBody() -> [String: JSONValue] {
        var body: [String: JSONValue] = [
            "runId": .string(runID),
            "sceneId": .string(sceneID),
            "renderer": .string(renderer.rawValue),
            "format": .string(format),
            "formatVersion": .number(Double(formatVersion)),
            "width": .number(Double(width)),
            "height": .number(Double(height)),
            "bytes": .number(Double(bytes)),
            "sha256": .string(sha256),
            "primitives": .number(Double(primitives)),
        ]
        if let preview { body["preview"] = .object(preview.jsonBody()) }
        return body
    }
}

public struct RenderReadyBody: Equatable, Codable, Sendable {
    public let runID: String
    public let sceneID: String
    public let renderer: RenderKind
    public let formatVersion: Int
    public let width: Int
    public let height: Int
    public let bytes: Int
    public let primitives: Int

    public init(runID: String, sceneID: String, prepare: RenderPrepareBody) {
        self.runID = runID
        self.sceneID = sceneID
        self.renderer = prepare.renderer
        self.formatVersion = prepare.formatVersion
        self.width = prepare.width
        self.height = prepare.height
        self.bytes = prepare.bytes
        self.primitives = prepare.primitives
    }

    public func jsonBody() -> [String: JSONValue] {
        [
            "runId": .string(runID),
            "sceneId": .string(sceneID),
            "renderer": .string(renderer.rawValue),
            "formatVersion": .number(Double(formatVersion)),
            "width": .number(Double(width)),
            "height": .number(Double(height)),
            "bytes": .number(Double(bytes)),
            "primitives": .number(Double(primitives)),
        ]
    }
}

public struct RenderRejectBody: Equatable, Codable, Sendable {
    public let runID: String
    public let sceneID: String
    public let code: RenderRejectCode

    public init(runID: String, sceneID: String, code: RenderRejectCode) {
        self.runID = runID
        self.sceneID = sceneID
        self.code = code
    }

    public func jsonBody() -> [String: JSONValue] {
        [
            "runId": .string(runID),
            "sceneId": .string(sceneID),
            "code": .string(code.rawValue),
        ]
    }
}

public struct RenderResultBody: Equatable, Codable, Sendable {
    public let runID: String
    public let sceneID: String
    public let renderer: RenderKind
    public let formatVersion: Int
    public let success: Bool
    public let bytes: Int
    public let primitives: Int
    public let validateMilliseconds: Int
    public let renderMilliseconds: Int
    public let errorCode: RenderRejectCode?

    public init(
        runID: String,
        sceneID: String,
        renderer: RenderKind,
        formatVersion: Int,
        success: Bool,
        bytes: Int,
        primitives: Int,
        validateMilliseconds: Int,
        renderMilliseconds: Int,
        errorCode: RenderRejectCode? = nil
    ) {
        self.runID = runID
        self.sceneID = sceneID
        self.renderer = renderer
        self.formatVersion = formatVersion
        self.success = success
        self.bytes = bytes
        self.primitives = primitives
        self.validateMilliseconds = validateMilliseconds
        self.renderMilliseconds = renderMilliseconds
        self.errorCode = errorCode
    }

    public func jsonBody() -> [String: JSONValue] {
        var body: [String: JSONValue] = [
            "runId": .string(runID),
            "sceneId": .string(sceneID),
            "renderer": .string(renderer.rawValue),
            "formatVersion": .number(Double(formatVersion)),
            "status": .string(success ? "ok" : "error"),
            "bytes": .number(Double(bytes)),
            "primitives": .number(Double(primitives)),
            "validateMs": .number(Double(validateMilliseconds)),
            "renderMs": .number(Double(renderMilliseconds)),
        ]
        if let errorCode {
            body["errorCode"] = .string(errorCode.rawValue)
        }
        return body
    }
}
